apps/validation.py

```diff
@@ -0,0 +1,153 @@
+            if new_gen is not None:
+                done = False
+                gen = new_gen
+                reset()
+                continue
+
+        try:
+            element = next(gen)
+        except StopIteration:
+            done = True
+            continue
+
+        yield element 
+
+@bear
+async def app(page):
+    q = Queue()
+    debounced = ClientWrap(q, debounce=0.3,form=True)
+    page["head"].print(
+        H.link(rel="stylesheet", href=here.parent / "paperoni" / "default.css")
+    )
+    area = H.div["area"]().autoid()
+    page.print(H.form(
+            H.input(name="title",placeholder="Title", oninput=debounced),
+            H.input(name="author",placeholder="Author", oninput=debounced),
+            H.br,
+            "Start Date",
+            H.input(type="date", id="start", name="date-start", oninput=debounced)["calender"],
+            H.br,
+            "End Date",
+            H.input(type="date", id="start", name="date-end", oninput=debounced)["calender"],
+            ))
+    page.print(area)
+    
+    def regen(event=None):
+        if event is not None:
+            title = event["title"]
+            author = event["author"]
+            date_start = event["date-start"]
+            date_end = event["date-end"]
+            return generate(title,author,date_start,date_end)
+        return generate()
+
+    def generate(title = None, author = None, date_start = None, date_end = None):
+        stmt = select(sch.Paper)
+        if not all(val is "" or val is None for val in [title,author,date_start,date_end]):
+            stmt = search(title,author,date_start,date_end)
+        try:
+            results = list(db.session.execute(stmt))
+            for (r,) in results:
+                yield r
+        except Exception as e: 
+            print("Error : ", e)
+        
+
+    def search(title,author,date_start,date_end):
+        stmt = select(sch.Paper)
+        #Selecting from the title
+        if title is not None and title != "":
+            stmt = select(sch.Paper).filter(sch.Paper.title.like(f"%{title}%"))
+        #Selecting from author
+        if author is not None and author != "":
+            stmt = (
+                    stmt.join(sch.Paper.paper_author).join(
+                        sch.PaperAuthor.author
+                    )
+                    .filter(sch.Author.name.like(f"%{author}%"))
+                )
+
+        #Selecting from date
+        if date_start is not None and date_start != "":
+            date_start_stamp = int(datetime(*map(int, date_start.split("-"))).timestamp())
+            stmt = stmt.join(sch.Paper.release).join(sch.Release.venue).filter(sch.Venue.date >= date_start_stamp)
+        if date_end is not None and date_end != "":
+            date_end_stamp = int(datetime(*map(int, date_end.split("-"))).timestamp())
+            stmt = stmt.join(sch.Paper.release).join(sch.Release.venue).filter(sch.Venue.date <= date_end_stamp)
```

Comment:
There is an error when both date_start and date_end are set because of a "column ambiguity". My guess is that there should not be two join statements: if there is both date_start and date_end, only call stmt.join().join() once, and then add two filters.

Reply:
Fixed in [ae336b8](https://github.com/mila-iqia/paperoni/pull/19/commits/ae336b87e8976be4102bd08decdc415d27a0f01e)